### PR TITLE
Fix #7861: Adds Query Parameter Filtering/Stripping

### LINKED ISSUE #7861: Add query param stripping
### Description: 
We need to implement query param stripping in iOS. Most of the logic is found in brave-core

### Test plan
1. Test all links on the following page: https://fmarier.github.io/brave-testing/query-filter.html
2. Test all links on the following page: https://fmarier.org/query-filter-test.html (It should also pass the tests that claim that "Don't work as expected"). In this page, trackers such as `fbclid` should be removed on all pages.
3. Make sure that back and forward works fine as well on some of these links. There shouldn't be any strangenesses with using back and forward buttons while navigating these links.

### PR DESCRIPTION
**Pull request by cuba**
## Summary of Changes
We want to add the query param stripping feature we have in desktop to iOS. Most of the changes are in brave-core

This pull request fixes #7861

Sec review
https://github.com/brave/reviews/issues/1368

Brave-core dependency:
https://github.com/brave/brave-browser/issues/32279

## Submitter Checklist:

- [x] *Unit Tests* are updated to cover new or changed functionality
- [x] User-facing strings use `NSLocalizableString()`
- [x] New or updated UI has been tested across:
  - [x] Light & dark mode
  - [x] Different size classes (iPhone, landscape, iPad)
  - [x] Different dynamic type sizes

## Test Plan:

STR in issue:

## Reviewer Checklist:

- [x] Issues include necessary QA labels:
  - `QA/(Yes|No)`
  - `bug` / `enhancement`
- [x] Necessary [security reviews](https://github.com/brave/security/issues/new/choose) have taken place.
- [x] Adequate unit test coverage exists to prevent regressions.
- [x] Adequate test plan exists for QA to validate (if applicable).
- [x] Issue and pull request is assigned to a milestone (should happen at merge time).

### CODE REVIEW

File: Sources/Brave/Frontend/Browser/BrowserViewController/BrowserViewController+WKNavigationDelegate.swift
Comment (Brandon-T) [inline Sources/Brave/Frontend/Browser/BrowserViewController/BrowserViewController+WKNavigationDelegate.swift]: Instead of returning the existing URL that was passed in, this should return `Optional<URL>` instead since the Brave-Core API can return `null`. Then you can just check if the result is null, and if it is, do nothing. If it isn't, redirect.

**Review by Brandon-T — CHANGES_REQUESTED** [review]

File: Sources/Brave/Frontend/Browser/BrowserViewController/BrowserViewController+WKNavigationDelegate.swift
Comment (cuba) [reply Sources/Brave/Frontend/Browser/BrowserViewController/BrowserViewController+WKNavigationDelegate.swift]: Sure that's a good idea.

Comment (cuba) [reply Sources/Brave/Frontend/Browser/BrowserViewController/BrowserViewController+WKNavigationDelegate.swift]: ok did this

**Comment by fmarier** [comment]
This PR has a lot of unrelated changes, does it need to be rebased?

**Comment by cuba** [comment]
> This PR has a lot of unrelated changes, does it need to be rebased?

sorry the parent branch changed and I needed to rebase. Its done now.

File: Sources/Brave/Frontend/Browser/BrowserViewController/BrowserViewController+WKNavigationDelegate.swift
Comment (fmarier) [inline Sources/Brave/Frontend/Browser/BrowserViewController/BrowserViewController+WKNavigationDelegate.swift]: A small optimization you can make here is to check if the `requestURL` has a query string before running through this code. We do this on desktop too: https://github.com/brave/brave-core/blob/7fb458e780ab6165f4a44402b4fe59cbd8647ff4/browser/net/brave_site_hacks_network_delegate_helper.cc#L90-L92

**Review by fmarier — COMMENTED** [review]
Will this respect the Shields setting for the page?

i.e. never strip query params for brave.com if brave.com is set to Shields down

**Review by Brandon-T — APPROVED** [review]

**Review by pes10k — APPROVED** [review]
LGTM (though keep in mind i am not really much of a swift reviewer)

**Comment by cuba** [comment]
> LGTM (though keep in mind i am not really much of a swift reviewer)

That's ok. You're just one of many reviewers that left comments and anyone that leves comments I hope to get an approval from based on their best knowledge.

**Review by soner-yuksel — CHANGES_REQUESTED** [review]
Please mention in description of the PR which PR needs to be merged on brave-core side to make this work. And write a proper Test Plan for QA.

And add **unit tests** for stripped URL functionality.

**Comment by cuba** [comment]
> Please mention in description of the PR which PR needs to be merged on brave-core side to make this work. And write a proper Test Plan for QA.
> 
> And add **unit tests** for stripped URL functionality.

The tests are in brave-core. Adding tests for the BVC changes is extremely tricky (I'm not even sure possible) as such, we don't have such BVC tests in iOS as of yet. I can certainly experiment but this will likely not happen on this PR.

**Comment by soner-yuksel** [comment]
> > Please mention in description of the PR which PR needs to be merged on brave-core side to make this work. And write a proper Test Plan for QA.
> > And add **unit tests** for stripped URL functionality.
> 
> The tests are in brave-core. Adding tests for the BVC changes is extremely tricky (I'm not even sure possible) as such, we don't have such BVC tests in iOS as of yet. I can certainly experiment but this will likely not happen on this PR.

It doesnt have to BVC direct tests here, just saying this can be nice to demonstrate `getInternalRedirect` or functionality inside is testable and results are expected.

I will check and approve when the core is merged and brave-core bump (the one that includes this change) is merged into develop.

File: Sources/Brave/Frontend/Browser/BrowserViewController/BrowserViewController+WKNavigationDelegate.swift
Comment (cuba) [reply Sources/Brave/Frontend/Browser/BrowserViewController/BrowserViewController+WKNavigationDelegate.swift]: This is done now on the brave core side

**Review by soner-yuksel — APPROVED** [review]

**Pull request merged**
